test/functional/test-element-stub.js

```diff
@@ -0,0 +1,156 @@
+    });
+  }
+
+  it('insert script', () => {
+    return getElementStub('amp-ad').then(() => {
+      resetAdScriptInserted();
+      expect(iframe.doc.querySelectorAll('amp-ad').length).to.equal(1);
+      expect(iframe.doc.head.querySelector('[custom-element="amp-ad"]'))
+        .to.be.null;
+      elementStub = new ElementStub(iframe.doc.body.firstChild.firstChild);
+      expect(iframe.doc.head.querySelector('[custom-element="amp-ad"]'))
+        .not.to.be.null;
+    });
+  });
+
+  it('only insert script once', () => {
+    return getElementStub('amp-ad').then(() => {
+      resetAdScriptInserted();
+      expect(iframe.doc.querySelectorAll('amp-ad').length).to.equal(1);
+      expect(iframe.doc.head.querySelector('[custom-element="amp-ad"]'))
+        .to.be.null;
+      elementStub1 = new ElementStub(iframe.doc.body.firstChild.firstChild);
+      expect(iframe.doc.head.querySelectorAll('[custom-element="amp-ad"]')
+        .length).to.equal(1);
+      elementStub2 = new ElementStub(iframe.doc.body.firstChild.firstChild);
+      expect(iframe.doc.head.querySelectorAll('[custom-element="amp-ad"]')
+        .length).to.equal(1);
+    });
+  });
+
+  it('insert script for amp-embed', () => {
+    return getElementStub('amp-embed').then(() => {
+      resetAdScriptInserted();
+      expect(iframe.doc.querySelectorAll('amp-embed').length).to.equal(1);
+      expect(iframe.doc.head.querySelector('[custom-element="amp-ad"]'))
+        .to.be.null;
+      expect(iframe.doc.head.querySelector('[custom-element="amp-embed"]'))
+        .to.be.null;
+      elementStub1 = new ElementStub(iframe.doc.body.firstChild.firstChild);
+      expect(iframe.doc.head.querySelector('[custom-element="amp-embed"]'))
+        .to.be.null;
+      expect(iframe.doc.head.querySelectorAll('[custom-element="amp-ad"]')
+        .length).to.equal(1);
+    });
+  });
+
+  it('should not insert when script exists in head', () => {
+    return getElementStub('amp-ad').then(() => {
+      resetAdScriptInserted();
+      const ampAdScript = iframe.doc.createElement('script');
+      ampAdScript.setAttribute('custom-element', 'amp-ad');
+      scriptSrc = 'http://localhost:8000/dist/v0/amp-ad-0.1.max.js';
+      expect(iframe.doc.head.querySelector('[custom-element="amp-ad"]'))
+        .to.be.null;
+      iframe.doc.head.appendChild(ampAdScript);
+      expect(iframe.doc.head.querySelectorAll('[custom-element="amp-ad"]')
+        .length).to.equal(1);
+      elementStub = new ElementStub(iframe.doc.body.firstChild.firstChild);
+      expect(iframe.doc.head.querySelectorAll('[custom-element="amp-ad"]')
+        .length).to.equal(1);
+    });
+  });
+
+  it('script should get correct attributes', () => {
+    return getElementStub('amp-ad').then(() => {
+      resetAdScriptInserted();
+      expect(iframe.doc.head.querySelector('[custom-element="amp-ad"]'))
+        .to.be.null;
+      elementStub = new ElementStub(iframe.doc.body.firstChild.firstChild);
```

Comment:
can you query select the #parent and then do firstChild, will make the test less brittle